Issue/5604

paulp opened the pull request:
Retargeting #1376 to master.

paulp committed:
Pending tests for SI-5954, SI-6225, SI-5877, SI-4695.

Which are all package object bugs.  Plus one more test
regarding package object overloading.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/571/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1279/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1279/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/571/

paulp commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -605,23 +605,31 @@ trait Typers extends Modes with Adaptations with Tags {
+     *  what other ways we can before pushing that way.
      */
     private def isInPackageObject(sym: Symbol, pkg: Symbol) = {
-      def isInPkgObj(sym: Symbol) =
```
I initially thought the problem was here, but I had been deceived by the improbable application of newlines.

paulp commented on src/reflect/scala/reflect/internal/TreeGen.scala:
```diff
@@ -172,10 +172,29 @@ abstract class TreeGen extends macros.TreeBuilder {
+        && qualsym.isPackage
+        && !sym.isDefinedInPackage
+      )
       val pkgQualifier =
```
This is the key change.  "sym.owner.isPackageObjectClass && sym.effectiveOwner == qual.tpe.typeSymbol" limits it to recognizing symbols which are defined in the package object itself, excluding any inherited symbols.

retronym replied:
One last question: Why did the bug only appear when compiling against the package object as a class files, but not as a source file?

paulp replied:
The Select tree must not come from mkAttributedSelect in that case.  It looks like makeAccessible does the same thing, but will not have the same bug because it uses "isInPackageObject".  Ah, so many code paths to approximately the same place.

There are also possibly ugly implications to the fact that isInPackageObject will give false negatives if either the package or package object symbol is not initialized.  I don't know if that ever happens.

paulp commented on src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala:
```diff
@@ -604,11 +604,18 @@ abstract class GenASM extends SubComponent with BytecodeWriters {
         val internalName = cachedJN.toString()
         val trackedSym = jsymbol(sym)
         reverseJavaName.get(internalName) match {
```
This is only a message cleanup, since I kept running into it during the development of this patch.

paulp committed:
Expanded an error message from the backend.

paulp committed:
Comments explaining some brokenness in Namers.

paulp committed:
Fixed SI-5604, selections on package objects.

mkAttributedSelect, which creates a Select tree based on
a symbol, has been a major source of package object bugs,
because it has not been accurately identifying selections
on package objects.  When selecting foo.bar, if foo turns
out to be a package object, the created Select tree must be

  foo.`package`.bar

However mkAttributedSelect was only examining the owner of
the symbol, which means it would work if the package object
defined bar directly, but not if it inherited it.

paulp committed:
Simplifications in typedIdent.

These accompany the changes to mkAttributedSelect in the
prior commit, and document additional brokenness with package
objects which stillr emains.

paulp commented:
@retronym, I divided this into smaller commits with better messages.  Assuming the tests pass, does it LG to Y?

paulp commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/625/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1334/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/625/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1334/

retronym commented:
Looks great, thanks for taking the time to tell the story page by page.

Package object flakiness is a horrible thing for users to encounter and minimize; this is a clear step forward. BTW, does this knock off SI-4695?

In fact, the only problem I have with this PR is the target branch. Can we request a second look from @gkossakowski ?

paulp commented:
It doesn't fix SI-4695 but it does pinpoint it.  See [this comment](https://github.com/paulp/scala/commit/97ede5a64a60c739f03619f9c0d7e2bf88a97207#L0R4818).

I agree about the wrong branch but I'm saving my energy for #1399.

ijuma commented:
Package object flakiness is indeed a horrible thing for users. Great to see Paul's efforts in this area.

gkossakowski commented:
This looks very good to me both comments and commit messages.

When it comes to target branch I disagree. That ship has sailed and let's accept that. By saturday me and @jsuereth will set up branches for 2.10.1 development and probably it can be back-ported there. Having it in master now will be good as well so merging.
